spec/index.bs

```diff
@@ -4898,12 +4898,14 @@ typedef double GPUPipelineConstantValue; // May represent WGSL's bool, f32, i32,
                     </dl>
                 ::
                     If |entry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/minBindingSize}} is not `0`:
-                    - If the last field of the corresponding structure defined in the shader has an unbounded array type,
+                    - If the store type of the corresponding buffer variable in the shader is a structure type, and the
+                        last member of that structure is an runtime-sized array,
                         then the value of |entry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/minBindingSize}}
-                        must be greater than or equal to the byte offset of that field plus the stride of the unbounded array.
-                    - If the corresponding shader structure doesn't end with an unbounded array type,
-                        then the value of |entry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/minBindingSize}}
-                        must be greater than or equal to the size of the structure.
+                        must be greater than or equal to the byte offset of that field plus the element stride of that array.
+                        That is, the binding size must be large enough to cover the whole structure including at least one
+                        element of that array.
+                    - Otherwise, the value of |entry|.{{GPUBindGroupLayoutEntry/buffer}}.{{GPUBufferBindingLayout/minBindingSize}}
```

Comment:
I think this is more nuanced than originally expected.
A struct with a single member has different layout than this single member on its own:
> SizeOf(S) = roundUp(AlignOf(S), OffsetOfMember(S, L) + SizeOfMember(S, L))

Consider `vec3<f32>`. The standalone size is 12 bytes, but the wrapping struct size is 16 bytes.

Hence, I believe an investigation is needed for all the backing APIs to figure out if they can support reduced binding sizes. This is *not* just a syntactic "convenience" feature on WGSL like I originally thought.

Reply:
This is where going a single path (i.e. `struct` only) benefits the spec. Because now we have to define the rules for non-struct types used in buffers specifically, test it. We'll then have a bunch of users complaining about this, since it is complicated. And our implementations would need to handle those cases carefully. So much for a "free" feature, cc @ben-clayton .

Reply:
@jrprice prodded me to look at this again.

First, I think the `vec3<f32>` case is the only problematic case.

Second, James pointed out that Dawn has end-to-end tests for the `struct  { vec3<f32> }`  case in a 12-byte buffer. It works on Metal, D3D12, and Vulkan.

I chased down the Vulkan side. Vulkan is ok with this, and in fact also ok with putting putting an array into a buffer that cuts off the padding at the end of the last element. (So if the array has N elements, stride K, and element with footprint size S (S < K of course), then the array can be put into a buffer with  (N-1) * K + S bytes.  ).
HLSL and D3D apparently also allow this stuff-extra-at-end-padding technique.

However, we kept things a bit simple for now, and had WGSL define structure and array sizes that were a little more strict than  what Vulkan needs (for example), and essentially disallow the stuff-extra-at-end-padding.
That's why the "size" of a structure is rounded up to the element alignments, and why array "size" is N * stride.
Aside from simplicity, it keeps things compatible with OpenGL type layouts.

So, in theory there are a few options:
1. Complexify the WGSL type size rules to allow us to cut corners and stuff things into end padding.    Then this changes the definition of "size" used by API-side binding validation. To me this is a big cost on the WGSL side to save a small change on the API side. 
2. Do what's proposed in this PR.  (And it still means Dawn is validating the wrong buffer size(!))
3. Change API-side binding validation so it uses the more relaxed size where we allow cutting off the padding at the end of the buffer variable store type.   Concretely, for `struct {vec3<f32>}` still define it to have a "size" of 16, but allow it to be mapped to a 12-byte buffer. This would require is to make a parallel set of "size" rules for WGSL types.   

I think 2 is the best option, by far. At least at this point.

Reply:
I guess there is also what James has been pushing me toward all along I guess:

4.  Ignore the fact that `vec3<f32>` turns into `struct { vec3<f32>; }` at the Vulkan level.  Vulkan is ok with 12 bytes for the latter, so don't force the *more strict* WGSL size to bleed into the validation rule at the WGSL/WebGPU level.  This means the Dawn end2end test case is ok, and actually justifies it.  

I guess I had to go dig into the Vulkan rules to realize this was totally ok.

It seems 4 is the winner?!

Reply:
Related to #425, https://github.com/austinEng/webgpu-samples/issues/149, and #1558
Current restriction is *not* for OpenGL, which is out of scope. Current restriction is for Metal, and I'm not sure why it worked for you. Are you sure you had Metal validation enabled? Are you sure the buffer you bound had exactly 12 bytes in it? I bet Dawn just rounded up the size of the buffer to 16 bytes, and that's why you weren't seeing the problem.

Reply:
Yes, I believe Dawn is indeed rounding the buffer size up to a multiple of 16 bytes, and that's a perfectly reasonable solution to this problem IMHO.

Reply:
Interesting.  We can buy the simplicity in the spec by doing the rounding-up-buffer-size in the browser code.

Anyway, I've updated to option 4.

Reply:
It reads to me that the Metal constraint also applies in other cases, e.g. `struct { i32; }` which have no interaction at all with alignments.  So you have to do something special for Metal anyway, both for sizes and alignments.